[CS-3929] Add cache invalidation once user request a card

paulinhapenedo opened the pull request:
### Description
This PR handles the use case when the user requests a card and goes back to the home screen. At this point, we're still not sure if the claim was successful or not, so we're adding a `pollInterval` to the claimed query based on the presence of a flag saved on the async storage.

- [x] Completes #CS-3929

### Checklist

- [x] Tested on a small device
- [x] Tested on iOS
- [x] Tested on Android

paulinhapenedo committed:
chore: add claimed to EoaClaimed interface

paulinhapenedo committed:
chore: add new key to async-utils file

paulinhapenedo committed:
feat: adds polling to the EoaClaimedQuery

oidoug commented on cardstack/src/components/CtaBanner/useWelcomeCtaBanner.ts:
```diff
@@ -15,14 +17,14 @@ const WELCOME_BANNER_KEY = 'WELCOME_BANNER_KEY';
-    },
-    { skip: !accountAddress }
+      skip: !accountAddress,
+      pollingInterval: hasClaimed ? undefined : 5000,
```
nit: extract the constant value to top of the file.

oidoug commented on cardstack/src/components/CtaBanner/useWelcomeCtaBanner.ts:
```diff
@@ -57,6 +59,36 @@ export const useWelcomeCtaBanner = () => {
 
+  const handleClaimStatus = useCallback(async () => {
+    try {
+      if (emailDropGetData?.claimed) {
```
Claimed will return true on all runs after first claim, so I think it isn't a good pattern to re-set it in storage every time. I think you want to first check if the key is saved or not.

oidoug commented on cardstack/src/components/CtaBanner/useWelcomeCtaBanner.ts:
```diff
@@ -57,6 +59,36 @@ export const useWelcomeCtaBanner = () => {
+    } catch (error) {
+      logger.log('[Async Storage] Error getting key', error);
+    }
+  }, [accountAddress, emailDropGetData]);
```
For cases like this one, I prefer to use a redux slice + redux-persist instead of the storage api directly (even though its only used here), just because of separation of concerns.

Or maybe you could extract this save/restore action into a model, so this logic isn't mixed here in the hook.

paulinhapenedo replied:
I prefer the second option since we seem to not be adding new things to redux (maybe I'm wrong on that 🤔)

oidoug reviewed (approved):
Looks good and functional, but I left some concerns that would be nice to address before merging.

paulinhapenedo committed:
refactor: move email card drop request keychain to its own file

oidoug commented on cardstack/src/components/CtaBanner/useWelcomeCtaBanner.ts:
```diff
@@ -57,6 +60,40 @@ export const useWelcomeCtaBanner = () => {
+        return;
+      }
+
+      // status changed to claim: true, disables polling
```
```suggestion
      // status changed to claimed: true, disables polling
```

oidoug reviewed (approved):
Good job :)
added a minor suggestion.

paulinhapenedo committed:
refactor: move card drop polling into redux

paulinhapenedo committed:
refactor: remove useCtaBanner in favor of using redux slice

paulinhapenedo commented:
@oidoug @danibonilha asking again for feedback because I changed a few things.
Dani agrees with you @oidoug that the async key needs to be handled by redux, so I moved both keys there and deleted the `useCtaBanner` hook

paulinhapenedo committed:
add welcomeBannerSlice to persistConfig and fix dismissedByUser assertion

Signed-off-by: Paula Penedo <paulinhapenedo@gmail.com>

oidoug commented on cardstack/src/redux/welcomeBanner.ts:
```diff
@@ -0,0 +1,42 @@
+  name: 'welcomeBanner',
+  initialState,
+  reducers: {
+    dismissBanner(state) {
```
Nice! I really prefer this approach, much cleaner and standardize than having a worker hook using async storage like `useCtaBanner` used. Thanks :)

oidoug commented on src/redux/store.ts:
```diff
@@ -16,7 +17,11 @@ const persistConfig = {
+  whitelist: [
+    primarySafeSliceName,
+    biometryToggleSliceName,
+    welcomeBannerSliceName,
```
👏

oidoug reviewed (approved):
Good job, thanks for refactoring useCtaBanner's hook stuff as well!

danibonilha reviewed (approved):
Nice! 🚀
